src/base/DelegateInputLabelMixin.js

```diff
@@ -31,18 +32,60 @@ export default function DelegateInputLabelMixin(Base) {
       this[setState]({ ariaLabel });
     }
 
+    // Forward ARIA labelledby as an aria-label to the input element.
+    get ariaLabelledby() {
+      return this[state].ariaLabelledby;
+    }
+    set ariaLabelledby(ariaLabelledby) {
+      if (this.getRootNode() !== null) {
+        // @ts-ignore
+        let labelNode = this.getRootNode().querySelector(
+          `#${ariaLabelledby}`
+        );
+
+        labelNode.setAttribute('aria-hidden', true);
+        const ariaLabel = labelNode.innerText;
+        this[setState]({ ariaLabel, ariaLabelledby });
+      }
+    }
+
     get [defaultState]() {
       return Object.assign(super[defaultState] || {}, {
         ariaLabel: "",
+        ariaLabelledby: "",
       });
     }
 
     [render](changed) {
       super[render](changed);
+      if (this[firstRender]) {
+        if (this.getRootNode() !== null) {
+          // @ts-ignore
+          const labelNode = this.getRootNode().querySelector(
+            `[for="${this.id}"]`
+          )
+          if (labelNode !== null) {
+            const ariaLabel = labelNode.innerText;
+            labelNode.setAttribute('aria-hidden', true);
+            this[setState]({ ariaLabel });
+          } else {
+            const labelNode = this.closest('label');
+            if (labelNode !== null) {
+              const ariaLabel = labelNode.innerText;
+              this[setState]({ ariaLabel });
+            }
+          }
+        }
+      }
 
       if (changed.ariaLabel) {
         const { ariaLabel } = this[state];
         this[inputDelegate].setAttribute("aria-label", ariaLabel);
+        this.removeAttribute('aria-label');
```

Comment:
Removing the `aria-label` attribute has an important side effect: it will invoke the component's `attributeChangedCallback` again, this time passing in a value of `null`.

At the moment, that side effect is masked by defensive logic in AttributeChangedCallback to avoid invoking a property setter if an attribute is changed during rendering. If/when this call to `removeAttribute` is moved out of the rendering phase (and into `rendered`, say), that defensive logic will no longer apply. In other words, when someone sets `aria-label` to "Email", it will invoke the `ariaLabel` property with the value of "Email", then eventually cause the `aria-label` attribute to removed, which will invoke the `ariaLabel` property with the value of null — which will remove the intended label.

We can't just ignore attempts to set `ariaLabel` to null, because that's a valid scenario. Someone might set a label on a component, then remove it later.

So we'll need to introduce a flag to distinguish between _our_ removing the attribute (after forwarding it to the inner input) and someone _else_ removing the attribute (because they don't want the input to be labelled any more). Such a flag is a bit tricky; I can walk through the mechanics.